test/runtests.jl

```diff
@@ -238,6 +238,31 @@ end
     end
 end
 
+# example of a struct with different fields and properties
+struct FieldProps{NT <: NamedTuple{(:a, :b)}}
+    components::NT
+end
+
+Base.propertynames(obj::FieldProps) = (:a, :b)
+Base.getproperty(obj::FieldProps, name::Symbol) = getproperty(getfield(obj, :components), name)
+ConstructionBase.constructorof(::Type{<:FieldProps}) = (a, b) -> FieldProps((a=a, b=b))
```

Comment:
```suggestion
function ConstructionBase.setproperties(obj::FieldProps, patch::NamedTuple)
    old_props = getproperties(obj)
    new_props = setproperties(old_props, patch)
    FieldProps(new_props)
end
```
This is a bit of an awkward point. The specification is that `setproperties` and `getproperties` match and that `constructorof` and `getfield.(Ref(obj), fieldnames(obj))` match. 
So this overload of `constructorof` is a bit of a hack.

Reply:
Ok you are right. I could have sworn we changed that.